### ISSUE TITLE
Undo (Ctrl-Z) after applying a style in Light Table may lead to incorrect thumbnails

### ISSUE THREAD

**Issue by chrisaga** [darktable-org/darktable#6734]
**Describe the bug**

When you apply a style in the Light Table and undo it by pressing `Ctrl-Z`, the thumbnail is reverted to the original thumbnail (as if it has just been imported) instead of the one just before the style was applied.
Same issue on multiple selection

**To Reproduce**

1. Go to 'Light Table'
2. Select a photo which has been visibly edited (cropped or rotated is good)
3. **Apply a style** (I tested with a style which adds a watermark)
4. See that the thumbnail is correctly updated
4. Press `Ctrl-Z` to "**undo**"
6. See that the **thumbnail is incorrectly updated** (it is not cropped or rotated anymore).
7. Go to 'Dark Room'
8. See that (thanks God) you haven't lost your history
9. Go back to 'Light Table'
10. See that the thumbnail is correct now.

**Expected behavior**

I would expect that the thumbnail would reflect the actual state of the photo after the undo.

**Screenshots**

NA

**Platform (please complete the following information):**
- Darktable Version: [3.2.1]
- OS: [ Ubuntu 20.04 with KDE 5 ]
- OpenCL not activated
- Graphics card : Cape Verde XT [Radeon HD 7770/8760 / R7 250X]

**Additional context**

- I tested on a new collection imported in DT 3.2.1
- Undo after adding a tag or editing metadata behaves as expected

**Comment by TurboGit** [comment]
I bet you have the mode overwrite activated for applying the style, so expected.

**Comment by chrisaga** [comment]
No, I have the append mode activated.
Overwrite would make little sense in this use case. Moreover, as I said, the history is still OK when I open the DarkRoom and the right thumbnail is regenerated when I switch back to the LightTable.
**The issue seems to be that the thumbnail generated after the *undo* is not synchronized with the actual state of the photo (i.e. the history)**.

I hope this helps.

N.B. I would expect *undo* to reverse to the previous state, not the original state event if the overwrite mode would have been activated. I just verified that it doesn't, and I think that there is another issue here since the state of the history before the style application is probably lost.

**Comment by TurboGit** [comment]
Reproduced. Fix coming.

**Comment by chrisaga** [comment]
This issue seems to be corrected in darktable 3.4 but the fix introduces a **new issue when there are multiple instances** of one module in the development pipe. I experimented it with two instances of exposure module and with two instances of color balance module.

If you replay the steps from my first message on a photo with multiple instances of a module in its pipe you get the following :
* After reverting the style application with `Ctrl-Z` (step 5).
* You can see that the thumbnail is regenerated (step 6) with more or less garbage.
* You go to the **Dark Room** (step 7) and see that your photo has turned into garbage.

To recover your photo you have to get back in the history stack before the first duplicate instance of whatever module is duplicated.

You can see in the active module stack (i.e.pipe) that only the first instance is present (but others are in the history stack).

Darktable throws this kind of error message to stderr : 
```
cannot get iop-order for exposure instance 1
[dt_ioppr_check_iop_order] history module not used but enabled!! exposure 1(2147483647) image 8747 (dt_dev_read_history_no_image end)
```
The xmp sidecar file seems to be OK but removing the photo and re-importing it with its xmp file gets the same garbage.
